packages/react-dialog/Spec.md

```diff
@@ -2,62 +2,303 @@
+| overlay  | _slot_                        | undefined | Dimmed background of dialog              |
+
+> ⚠️ _Pending issue: `draggable`, the dragging functionality of the dialog is under consideration. This might be excluded from the dialog implementation._
+
+- `type` property (dialog variations):
+
+  - `modal`: When this type of dialog is open, the rest of the page is dimmed out and cannot be interacted with. The tab sequence is kept within the dialog and moving the focus outside the dialog will imply closing it. This is the default type of the component.
+
+  - `non-modal`: When a non-modal dialog is open, the rest of the page is not dimmed out and users can interact with the rest of the page. This also implies that the tab focus can move outside the dialog when it reaches the last focusable element.
+
+  - `alert`: is a special type of modal dialogs that interrupts the user's workflow to communicate an important message or ask for a decision. These dialogs are not dismissable, neither by escape key or by clicking outside the dialog.
+
+### DOM
+
+```html
+<div role="dialog" class="fui-dialog">{children}</div>
+```
+
+## Dialog Header
+
+The Header component will expect to have a dialog title/header and will show by default the close (X icon) button.
+
+### DOM
+
+```html
+<div class="fui-dialog-header">
+  <span>{title}</span>
+  <button aria-label="close" />
+</div>
+```
+
+### API
 
-_Provide some representative example code that uses the proposed API for the component_
+| Property          | Values  | Default     | Purpose                         |
+| ----------------- | ------- | ----------- | ------------------------------- |
+| `closeButton`     | _slot_  | `undefined` | Button that closes the dialog   |
+| `hideCloseButton` | boolean | `false`     | Shows or hides the close button |
 
-## Variants
+## Dialog Body
 
-_Describe visual or functional variants of this control, if applicable. For example, a slider could have a 2D variant._
+The body is a container where the content of the dialog is rendered. Apart from padding, this component does not have other behaviour.
 
-## API
+```html
+<div class="fui-dialog-body">{children}</div>
+```
 
-_List the **Props** and **Slots** proposed for the component. Ideally this would just be a link to the component's `.types.ts` file_
+## Dialog Footer
 
-## Structure
+The footer is a container for the actions of the dialog, which must be not more than 3 (primary, secondary and tertiary actions).
 
-- _**Public**_
-- _**Internal**_
-- _**DOM** - how the component will be rendered as HTML elements_
+### DOM
+
+```html
+<div class="fui-dialog-footer">
+  <button>{tertiaryAction}</button>
+  <button>{primaryAction}</button>
+  <button>{secondaryAction}</button>
+</div>
+```
+
+> ⚠️ _Pending issue: should there be any handling for the order of buttons? This also, includes the focus sequence of the buttons when the dialog has a tertiary button as well._
+
+## Sample Code
+
+### App
+
+```tsx
```

Comment:
Missing trigger pattern has some consequences. ⬇️

### No uncontrolled mode

Without triggers we can't really implement [uncontrolled mode](https://reactjs.org/docs/uncontrolled-components.html) as we don't control the state from a component anyhow. This might not be a big problem, but it diverges API significantly from other components.

### Does not prevent wrong usages by design

```tsx
function App() {
  return (
    <>
      {open && <Dialog />
      <Dialog open={open} />
    </>
  )
}
```

It's a common thing to use conditional rendering with React. On the first look there is nothing wrong with the snippet, results will be same. But once we will implement animations a `Dialog` that uses conditional rendering will never have proper closing animation.

It's a React problem with its lifecycle methods, but once a component gets unmounted it's not in DOM anymore.

How it's different from other components?

```tsx
function App() {
  // 🔴 `button` will not be rendered in this case
  return open ? <Tooltip><button /></Tooltip> : null
}
```

### Focus restoration issues

@ling1726 has already mentioned this in detail.

---

Overall, I think that React Spectrum has a nice contract for this case:
https://react-spectrum.adobe.com/react-spectrum/Dialog.html

```tsx
function App() {
  return (
    <>
      <DialogTrigger>
        <Button>Publish</Button>
        {(close) => (
          <Dialog>
            <Button onClick={close}>Cancel</Button>
          </Dialog>
        )}
      </DialogTrigger>

      <DialogTrigger>
        <Button>Publish</Button>
        <Dialog />
      </DialogTrigger>
    </>
  );
}
```

Which in our case will be transformed to:

```tsx
function App() {
  return (
    <Dialog>
      <DialogTrigger>
        <Button>Publish</Button>
      </DialogTrigger>
      {(close) => (
        <DialogContent>
          <Button onClick={close}>Cancel</Button>
        </DialogContent>
      )}
    </Dialog>
  );
}
```

This still diverges from other components, but solves issues mentioned above.
Please share your thoughts as might be reasons to follow API proposed in the RFC.

Reply:
My reasoning of not including the trigger pattern was to keep dialog as simple as possible without many containers. However, after discussing this with @ling1726 and @Hotell, it totally makes sense to have this, working on it already and I'll be adding it on the spec soon. Thanks 🙏